docs/doc/90-contributing/08-how-to-write-scalar-functions.md

```diff
@@ -65,161 +60,186 @@ This is very common optimization and widely used in arrow's compute system.
 
 ### Special column
 
--  Constant column
+- Constant column
 
-    Sometimes column is constant in the block, such as: `SELECT 3 from table`, the column 3 is always 3, so we can use a constant column to represent it. This is useful to save the memory space during computation.
+  Sometimes column is constant in the block, such as: `SELECT 3 from table`, the column 3 is always 3, so we can use a constant column to represent it. This is useful to save the memory space during computation.
 
-    So databend's DataColumn is represented by:
-
-    ```rust
-    #[derive(Clone)]
-    pub struct ConstColumn {
-        length: usize,
-        column: ColumnRef,
-    }
-    ```
 - Nullable column
 
-    By default, columns are not nullable. If we want a nullable column, we can use this to represent it.
+  By default, columns are not nullable. If we want a nullable column, we can use this to represent it.
 
-    ```rust
-    #[derive(Clone)]
-    pub struct NullableColumn {
-        validity: Bitmap,
-        column: ColumnRef,
-    }
-    ```
+## Function Registration
 
-### Writing function guidelines
+The `FunctionRegistry` is responsible for registering functions.
 
-## ScalarFunction trait introduction
+```rust
+#[derive(Default)]
+pub struct FunctionRegistry {
+    pub funcs: HashMap<&'static str, Vec<Arc<Function>>>,
+    #[allow(clippy::type_complexity)]
+    pub factories: HashMap<
+        &'static str,
+        Vec<Box<dyn Fn(&[usize], &[DataType]) -> Option<Arc<Function>> + 'static>>,
+    >,
+    pub aliases: HashMap<&'static str, &'static str>,
+}
+```
 
-All scalar functions implement `Function` trait, and we register them into a global static `FunctionFactory`, the factory is just an index map and the key is the name of the scalar function.
+It contains three HashMaps: `funcs`, `factories`, and `aliases`.
 
-:::tip
-    Function name in Databend is case-insensitive.
-:::
+`funcs` and `factories` store registered functions. The former registering functions with a fixed number of parameters (currently the minimum number of arguments is `0` and the maximum number of arguments is `5`), `register_0_arg`, `register_1_arg` and so on. The latter registering functions with variable-length parameter (such as concat) and call the `register_function_factory function`.
 
-``` rust
+`aliases` is used to store aliases for functions. A function may have more than one alias (e.g. `minus` has `subtract` and `neg` aliases), its key is the alias of a function, value is the name of the currently existing function, and the `register_aliases` function will be called.
```

Comment:
`aliases` uses key-value pairs to store aliases for functions. A function can have more than one alias (for example, `minus` has `subtract` and 'neg'). The key is the alias of a function, and the value is the name of the current function, and the `register_aliases` function will be called.